hail/python/hail/matrixtable.py

```diff
@@ -2871,38 +3060,43 @@ def show(
         def estimate_size(struct_expression):
             return sum(max(len(f), len(str(x.dtype))) + 3 for f, x in struct_expression.flatten().items())
 
-        if n_cols is None:
+        if n_rows is None:  # Careful with truthiness here, n_rows can be 0
+            n_rows = 10
+        n_rows = min(max(n_rows, 0), self.count_rows())
+        if n_cols is None or width is None:
             import shutil
 
-            (characters, _) = shutil.get_terminal_size((80, 10))
-            characters -= 6  # borders
-            key_characters = estimate_size(self.row_key)
-            characters -= key_characters
-            if include_row_fields:
-                characters -= estimate_size(self.row_value)
-            characters = max(characters, 0)
-            n_cols = characters // (estimate_size(self.entry) + 4)  # 4 for the column index
-        actual_n_cols = self.count_cols()
-        displayed_n_cols = min(actual_n_cols, n_cols)
-
-        t = self.localize_entries('entries', 'cols')
-        if len(t.key) > 0:
-            t = t.order_by(*t.key)
-        col_key_type = self.col_key.dtype
-
-        col_headers = [f'<col {i}>' for i in range(0, displayed_n_cols)]
-        if len(col_key_type) == 1 and col_key_type[0] in (hl.tstr, hl.tint32, hl.tint64):
-            cols = self.col_key[0].take(displayed_n_cols)
-            if len(set(cols)) == len(cols):
-                col_headers = [repr(c) for c in cols]
-
-        entries = {col_headers[i]: t.entries[i] for i in range(0, displayed_n_cols)}
-        t = t.select(
-            **{f: t[f] for f in self.row_key}, **{f: t[f] for f in self.row_value if include_row_fields}, **entries
-        )
+            (characters, _) = shutil.get_terminal_size()
+            if width is None:
+                width = characters
+            if n_cols is None:
+                characters -= 6  # borders
+                key_characters = estimate_size(self.row_key)
+                characters -= key_characters
+                if include_row_fields:
+                    characters -= estimate_size(self.row_value)
+                characters = max(characters, 0)
+                n_cols = characters // (estimate_size(self.entry) + 4)  # 4 for the column index
```

Comment:
So, one thing that has long confused users is that a Matrix Table with enough entry fields on a skinny enough terminal will show with zero columns. We do print the message about the number of columns, but users still get confused. Since you're changing this code anyway, can we add a `n_cols = max(n_cols, 1)` to ensure we always print at least one column?